pipeline/id_minter/src/test/scala/weco/pipeline/id_minter/steps/IdentifierGeneratorTest.scala

```diff
@@ -90,6 +96,114 @@ class IdentifierGeneratorTest
+        }.map(Identifier(i)).list.apply()
+
+        maybeIdentifiers should have length 2
+        maybeIdentifiers should contain theSameElementsAs
+          List(sourceIdentifier1, sourceIdentifier2).flatMap(ids.get)
+    }
+  }
+
+  it("tries looking up identifiers again if it fails to insert 'new' ids") {
+    // Simulation of a race condition.  This test represents the behaviour of the
+    // "second" participant in the race, with mocks representing the  "first"
+    // participant.
+    // When running in parallel (multiple threads, multiple hosts, whatever),
+    // it is possible for a clash to occur, where two processes create a new
+    // id for the same sourceIdentifier almost simultaneously.
+    // The second process requests the id from the database, but it is not
+    // yet there (because the first process has not yet stored it).
+    // The first process finishes storing the id, then the second process
+    // tries to store its new id, and fails.
+    // In this scenario, the second process goes back to the database to
+    // fetch the updated ids.
+    val sourceIdentifiers = (1 to 5).map(_ => createSourceIdentifier).toList
+    val canonicalIds = (1 to 5).map(_ => createCanonicalId).toList
+    val existingEntries = (sourceIdentifiers zip canonicalIds).map {
+      case (sourceId, canonicalId) =>
+        (
+          sourceId,
+          Identifier(
+            canonicalId,
+            sourceId
+          ))
+    }.toMap
+
+    val daoStub = mock[IdentifiersDao]("dao")
+
+    // On the first call, the DAO returns an empty set of found identifiers,
+    // and a full set of unminted identifiers, because the other participant
+    // in the race has not yet saved its identifiers.
+    // On the second call, they have been saved, so the full/empty states are switched.
+    when(daoStub.lookupIds(any[List[SourceIdentifier]])(any[DBSession]))
+      .thenReturn(
+        Success(
+          IdentifiersDao.LookupResult(
+            existingIdentifiers = Map.empty,
+            unmintedIdentifiers = sourceIdentifiers
+          )))
+      .thenReturn(
+        Success(
+          IdentifiersDao.LookupResult(
+            existingIdentifiers = existingEntries,
+            unmintedIdentifiers = Nil
+          )
+        ))
+
+    // When attempting to save, the first call fails, because the other participant
+    // has now saved its identifiers.
+    // The second run through will not call saveIdentifiers, as it will have retrieved
+    // all of the ids from the database.
+    when(daoStub.saveIdentifiers(any[List[Identifier]])(any[DBSession]))
+      .thenReturn(
+        Failure(IdentifiersDao.InsertError(Nil, new Exception("no."), Nil))
+      )
+
+    withIdentifierGenerator(Some(daoStub)) {
+      case (identifierGenerator, _) =>
+        val triedIds = identifierGenerator.retrieveOrGenerateCanonicalIds(
+          sourceIdentifiers
+        )
+
+        triedIds shouldBe a[Success[_]]
```

Comment:
Is it worth asserting that we retrieve the correct identifiers?

Reply:
Yes.  Originally. I thought that making that meaningful would overcomplicate the test,  but I realise that on the first attempt, it will generate new ids, so we need to ensure that they don't get returned.